File: 15反转链表返回表头/Solution.java
public class Solution {
    public ListNode ReverseList(ListNode head) {
         ListNode newList = new ListNode(-1);
    while (head != null) {
        ListNode next = head.next;
        head.next = newList.next;
        newList.next = head;
        head = next;
    }
    return newList.next;
    }
}
/*每次将q节点插入首个节点p的前面。为了进行下次操作，我们还需要在插入前记录下q后面节点t。

当q为空时，该方法结束，此时完成了整个链表的反转。最后，记得将头结点head指向反转后的首个节点p，

同时将反转后的尾节点的next指针设为空。*/

File: 15反转链表返回表头/Solution.cpp
class Solution {
public:
    ListNode* ReverseList(ListNode* head) {
           if (head==NULL||head->next==NULL) return head;
        ListNode* newhead=ReverseList(head->next);
        head->next->next=head;
        head->next=NULL;
        return newhead;
    }
};
/* 递归解决，首先通过递归到达链表的末尾，然后将末尾的前一个节点的next->next
也就是末尾的next指向末尾的前一个节点，通过反复的递归就可以将链表完全反转
但是注意在整个递归的过程中newhead一直都是链表的末尾 */